pkg/urbit/worker/main.c

```diff
@@ -567,17 +568,57 @@ _worker_work_live(c3_d evt_d, u3_noun job)
     u3x_trel(gon, 0, &vir, &cor);
 
     u3k(ovo); u3k(vir); u3k(cor);
-    u3z(gon); u3z(job);
+    u3z(gon); u3z(job); u3z(last_date);
 
-    _worker_sure(ovo, vir, cor);
+    u3V.dun_d = u3V.sen_d;
 
-    //  reclaim memory from persistent caches periodically
+    //  after a successful event, we check for memory pressure.
+    //
+    //    if we've exceeded either of two thresholds, we reclaim
+    //    from our persistent caches, and notify the daemon
+    //    (via a "fake" effect) that arvo should trim state
+    //    (trusting that the daemon will enqueue an appropriate event).
+    //    For future flexibility, the urgency of the notification is represented
+    //    by a *decreasing* number: 0 is maximally urgent, 1 less so, &c.
     //
-    //    XX this is a hack to work around the fact that
-    //    the bytecode caches grow rapidly and are not
-    //    able to be simply capped (due to internal posts).
+    //    high-priority: 2^22 contiguous words remaining (~8 MB)
+    //    low-priority:  2^27 contiguous words remaining (~536 MB)
+    //    XX maybe use 2^23 (~16 MB) and 2^26 (~268 MB?
     //
-    if ( 0 == (evt_d % 1000ULL) ) {
+    //    XX refactor: we should measure memory after losing the old kernel
+    //
+    {
+      u3_noun pri = u3_none;
+      c3_w pos_w = u3a_open(u3R);
+      c3_w low_w = (1 << 27);
+      c3_w hig_w = (1 << 22);
+
+      if ( (pre_w > low_w) && !(pos_w > low_w) ) {
```

Comment:
Might be nice to make a little printf here, or in arvo or somewhere that says something like "reclaiming space".  Or maybe only if `|verb` is on or something.  Maybe that's just something dill should do with @belisarius222's new `|knob`.